certbot-apache/certbot_apache/parser.py

```diff
@@ -303,21 +303,34 @@ def get_ifmod(self, aug_conf_path, mod, beginning=False, force_create=False):
         """
         if_mods = self.aug.match(("%s/IfModule/*[self::arg='%s']" %
                                   (aug_conf_path, mod)))
-        if len(if_mods) == 0 or force_create:
-            if beginning:
-                c_path_arg = "{}/IfModule[1]/arg".format(aug_conf_path)
-                # Insert IfModule before the first directive
-                self.aug.insert("{}/directive[1]".format(aug_conf_path),
-                                "IfModule", True)
-            else:
-                c_path = "{}/IfModule[last() + 1]".format(aug_conf_path)
-                c_path_arg = "{}/IfModule[last()]/arg".format(aug_conf_path)
-                self.aug.set(c_path, "")
-            self.aug.set(c_path_arg, mod)
-            if_mods = self.aug.match(("%s/IfModule/*[self::arg='%s']" %
-                                      (aug_conf_path, mod)))
+        if not if_mods:
+            return self.create_ifmod(aug_conf_path, mod, beginning)
+
         # Strip off "arg" at end of first ifmod path
-        return if_mods[0][:len(if_mods[0]) - 3]
+        return if_mods[0].rpartition("arg")[0]
+
+    def create_ifmod(self, aug_conf_path, mod, beginning=False):
+        """Creates a new <IfMod mod> and returs it path.
+
+        :param str aug_conf_path: Augeas configuration path
+        :param str mod: module ie. mod_ssl.c
+        :param bool beginning: If the IfModule should be created to the beginning
+            of augeas path DOM tree.
+
+        """
+        if beginning:
+            c_path_arg = "{}/IfModule[1]/arg".format(aug_conf_path)
+            # Insert IfModule before the first directive
+            self.aug.insert("{}/directive[1]".format(aug_conf_path),
+                            "IfModule", True)
+            retpath = "{}/IfModule[1]/".format(aug_conf_path)
+        else:
+            c_path = "{}/IfModule[last() + 1]".format(aug_conf_path)
+            c_path_arg = "{}/IfModule[last()]/arg".format(aug_conf_path)
+            self.aug.set(c_path, "")
+            retpath = "{}/IfModule[last()]/".format(aug_conf_path)
```

Comment:
Is there a way to normalize an Augeas path?

In the current PR we compare two Augeas paths to see if they are the same, but will `match` ever return a path using `last()` like this? It's probably not an issue for our current PR, but I'm curious if this change to the return value to `get_ifmod` will cause any problems anywhere else in the code base where we may try to compare paths like this.

Reply:
Match never returns a `last()` or any other dynamic XPATH value. Currently the get_ifmod is only used in `parser.add_dir_to_ifmodssl()` (in addition to the overrride_centos.py) and that doesn't compare paths, and while we compare paths in the CentOS override, the paths are always added as `beginning=True`, resulting to index 1. 

Previously this method added the new directive, then did a search, and returned the first value of the matches. I see this previous functionality potentially causing more issues than the current approach. We could figure our way around this by returning the last object in `if_mods` list of matches if `beginning=False`, and the first one otherwise. That said, I'm not 100% convenient that Augeas always returns the paths in the same order (re: spurious Apache test failures in VirtualHost matching).

I think that we are fine with this current approach, after I document the return value (as requested in the next review comment)

Reply:
That all makes sense. Thanks for diving into it a bit more to prove there's no problems caused by the changes here.

I don't think we need to do anything else now, but this is an ugly problem. It's not completely unrelated to the problems in #6256 which also is due to problems trying to use xpaths.

It's not on our roadmap or part of our OTF grant, but if we end up sticking with Augeas, I personally think we should take the time to do some serious refactoring of the code here as soon as we can. It feels like we're repeatedly hitting issues like this which I think we can solve by rethinking how we use Augeas and limiting its scope to a subset of the Apache plugin rather than letting it propagate everywhere. That will be a lot of work, but I feel like we're spending a lot of time fighting these issues on any significant Apache PR.

Do you agree? Feel free to say no, I'm just genuinely curious if you are feeling and thinking the same way about this.

Reply:
I absolutely and wholeheartedly agree, we had a discussion about this offline, and it looks like a good way forward in all aspects - also those relating to other development goals we have in our roadmap.